### DISCUSSION TITLE
Avalonia.Controls Packages

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#8008] [Ideas]
This originated as a counter-proposal to the main ideas in: "Discussion to create Avalonia.Toolkit" #7929

---

I'm generally against several of the ideas in the Toolkit discussion related to separating controls out of Avalonia itself. While the toolkit proposal is delayed for now, I know this discussion is going to come up again after 11.0 release (possible for 12.0). Therefore, I would like to offer some ideas and a general counter-proposal as a compromise.

My very general point (among others I'll save for discussion in #7929 itself) is that Controls are fundamentally important for a UI framework and shouldn't be in a more general-purpose toolkit. Controls are critically important to adoption of a UI framework and we've all come to expect a certain baseline of conrols from WPF/WinUI that I think Avalonia should support by default. However, certain Avalonia platforms and applications don't need anything but the very basics (Panel, etc.) anything more is overkill and can hurt performance. I think this, along with maintainability concerns, is the main reason the Toolkit was proposed.

So this counter-proposal is essentially creating separate nuget packages for controls; however, not putting them in a Toolkit. I think a Toolkit should be reserved for much more esoteric or experiental controls and also for non-control code (think MVVM stuff and helpers in the UWP community toolkit).

Here are some of the packages I'm proposing:
 1. `Avalonia.Controls` : Contains literally all the controls that can be separated from core Avalonia. If you are using Avalonia just for rendering or 2D/3D graphics you don't need this package. However, everyone else needs it and it should be included in all default project templates. Only those that don't need it and want to save space can remove it (opt-out). Note that controls in this package use the exact same `Avalonia.Controls` namespace so usage is transparent (sharing the same namespace is a key feature of this proposal).
 2. `Avalonia.Controls.DataGrid` : This comes along as it always has. It is installed by default with project templates and also is opt-out for those that don't need it and want to save space.
 3. `Avalonia.Controls.ColorPicker` : Will be a new package handled the same way as DataGrid. Included by default and opt-out for those that don't need it.

Future packages can easily be added.

This proposal's main points are:
 * Always put controls in the `Avalonia.Controls` or `Avalonia.Controls.*` namespace to avoid using/namespace trouble. It makes things transparent and appear fully integrated for the majority of developers.
 * Include all control packages by default in new project templates -- all of them! This gives developers a full-featured UI framework with all controls as the default experience much like WPF/WinUI. If they don't need something they can remove a package to save space (advanced/specialized use cases only).
 * Only large controls should go in a separate nuget package (DataGrid, ColorPicker, TreeDataGrid, NavigationView, etc.). This is a compromise with the core team and also helps ensure app sizes stay smaller if they never use this code. All basic controls, including the button variants, should go inside `Avalonia.Controls`.
 * These proposals, unlike the Toolkit, are fully transparent to developers. Only those that really need to optimize for storage/performance will need to remove a package. But it's simply opt-out.
 * It is probably not possible to implement this proposal until some of the 11.0 styling changes are complete.

**Comment by timunie** [comment]
Good idea imo. But one could have 2 or three templates then. I think of

- Minimal - no controls
- Basic - most basic controls only
- Full - all possible controls for opting out

**Reply by robloo** [reply]
Well, a few key points were:

1. Simplify things as if all is in core Avalonia and make it transparent for devs. Making new devs choose a template based on controls is not a simplification and will complicate things.
2. All controls should be provided by default but its possible to opt-out for advanced or specialized developers.
